crates/runtime/src/embeddings/execution_plan.rs

```diff
@@ -400,6 +407,75 @@ pub(super) async fn get_vectors(
     Ok(builder.finish())
 }
 
+/// Embed a [`StringArray`] using the provided [`Embed`] model with parallel processing.
+/// Similar to [`get_vectors`] but runs synchronously and processes embeddings in parallel
+/// across multiple threads using [`rayon::par_iter`]. The output is a [`FixedSizeListArray`],
+/// where each [`String`] gets embedded into a single [`f32`] vector.
+#[allow(clippy::cast_sign_loss, clippy::cast_possible_truncation)]
+pub(super) fn get_vectors_in_process<'a>(
+    arr: impl Iterator<Item = Option<&'a str>>,
+    model: &dyn Embed,
+    vector_length: i32,
+) -> Result<FixedSizeListArray, Box<dyn std::error::Error + Send + Sync>> {
+    let inputs: Vec<Option<String>> = arr.map(|o| o.map(ToString::to_string)).collect();
+
+    let mut builder = FixedSizeListBuilder::with_capacity(
+        PrimitiveBuilder::<Float32Type>::with_capacity(inputs.len() * (vector_length as usize)),
+        vector_length,
+        inputs.len(),
+    )
+    .with_field(Arc::new(Field::new("item", DataType::Float32, true)));
+
+    let pool = build_embedding_pool(model.parallelism())?;
+
+    // Check for null rows: embed 'string-at-a-time' if there are any, otherwise
+    // chunk embedding tasks into batches and use [`EmbeddingInput::StringArray`]
+    if inputs
+        .iter()
+        .any(|o| !matches!(o, Some(s) if !s.is_empty()))
+    {
+        let embeds: Vec<_> = pool.install(|| {
+            inputs
+                .into_par_iter()
+                .map(|o| match o {
+                    Some(input) if input.is_empty() => None,
+                    Some(input) => model.embed_sync(EmbeddingInput::String(input)).ok(),
+                    _ => None,
+                })
+                .collect()
+        });
+
+        for embed in embeds {
+            if let Some(embedding) = embed {
+                builder.values().append_slice(&embedding[0]);
+                builder.append(true);
+            } else {
+                builder.values().append_nulls(vector_length as usize);
```

Comment:
```suggestion
```

You shouldn't need to append nulls to the values array, marking the list array as null will set the correct offsets such that the next non-null list item will start from the correct offset, i.e.:

```markdown
+---------------------+---------------------------+
| List Validity Bitmap| Values Array (Float32)    |
+---------------------+---------------------------+
| [true, false, true] | [1.0, 2.0, 3.0,           |  <- Offset 0: N=0
|                     |  4.0, 5.0, 6.0]           |  <- Offset 3: N=2
+---------------------+---------------------------+
```

Reply:
ah maybe fixed sized lists work differently?